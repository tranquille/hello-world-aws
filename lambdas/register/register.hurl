POST https://tapam2fabj.execute-api.eu-central-1.amazonaws.com/stage/register
{
    "email": "knut@asdf.de",
    "password": "qwert1Z",
    "sports": ["Basketball", "Vollyball"]

}
