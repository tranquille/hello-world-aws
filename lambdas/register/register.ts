import { APIGatewayEvent } from 'aws-lambda';

export const handler = async(event: APIGatewayEvent) => {
    try {
        return {
          status: 200,
          body: JSON.stringify({"hello": "world"}),
          headers: {
            "content-type": "application/json"
          }
        };
    }
    catch (e) {
        console.error(e);
        return 500;
    }
};