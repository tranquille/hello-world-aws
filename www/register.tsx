import { Accessor, Setter, createSignal } from "solid-js";
import { register } from "./api-client";
import { useNavigate } from "@solidjs/router";

export default function Register() {
  const [sports, setSports]: [
    sports: Accessor<Array<string>>,
    setSports: Setter<Array<string>>
  ] = createSignal([]);
  const [email, setEmail] = createSignal("");
  const [password, setPassword] = createSignal("");
  const [hasError, setError] = createSignal("");

  const navigate = useNavigate();

  const handleSports = (e: string) => {
    const s = e.split(",");
    setSports(s);
  };

  const handleSubmit = async () => {
    setError("");
    try {
      console.log(sports());
      const registerResult = await register(email(), password(), sports());

      if (registerResult.name) {
        switch (registerResult.name) {
          case "UsernameExistsException":
            setError("Username exists");
            break;
          case "InvalidPasswordException":
            setError("Password is invalid");
            break;
          default:
            setError("Registration failed...");
        }
        return;
      }

      navigate("/", { replace: true });
    } catch (error: unknown) {
      setError("Regestration failed...");
      console.log("asdsad");
      console.error(error);
    }
  };

  return (
    <>
      <h2>Register</h2>
      <div>
        <input
          name="username"
          type="text"
          placeholder="Username"
          value={email()}
          onInput={(e) => setEmail(e.target.value)}
          required
        />
      </div>
      <div>
        <input
          name="password"
          type="password"
          placeholder="Password"
          value={password()}
          onInput={(e) => setPassword(e.target.value)}
          required
        />
      </div>
      <div>
        <input
          name="sports"
          type="text"
          placeholder="Sports - comma separated"
          onInput={(e) => handleSports(e.target.value)}
          required
        />
      </div>
      <div>
        <button onClick={() => handleSubmit()}>Submit</button>
      </div>
      {hasError() && <span>{hasError()}</span>}
    </>
  );
}
