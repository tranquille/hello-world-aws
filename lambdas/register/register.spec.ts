vi.mock("@aws-sdk/util-dynamodb");

const identityProviderSend = vi.fn();
vi.mock("@aws-sdk/client-cognito-identity-provider", async () => {
  // rome-ignore lint/suspicious/noExplicitAny: <explanation>
  const actual: any = await vi.importActual(
    "@aws-sdk/client-cognito-identity-provider",
  );

  const CognitoIdentityProviderClient = vi.fn(() => ({
    send: identityProviderSend,
  }));

  return { ...actual, CognitoIdentityProviderClient };
});

const dynamoDbSend = vi.fn();
vi.mock("@aws-sdk/client-dynamodb", async () => {
  // rome-ignore lint/suspicious/noExplicitAny: <explanation>
  const actual: any = await vi.importActual("@aws-sdk/client-dynamodb");

  const DynamoDBClient = vi.fn(() => ({
    send: dynamoDbSend,
  }));

  return { ...actual, DynamoDBClient };
});

import { APIGatewayProxyEvent } from "aws-lambda";
import { handler } from "./register";

const getTestEvent = (_payload: unknown, isBase64Encoded = false) => {
  let payload = JSON.stringify(_payload);
  if (isBase64Encoded) {
    payload = Buffer.from(JSON.stringify(_payload)).toString("base64");
  }

  return {
    body: payload,
    headers: {},
    multiValueHeaders: {},
    httpMethod: "POST",
    isBase64Encoded,
    path: "",
    pathParameters: null,
    queryStringParameters: null,
    multiValueQueryStringParameters: null,
    stageVariables: null,
    // rome-ignore lint/suspicious/noExplicitAny: It's a test
    requestContext: {} as any,
    resource: "",
  };
};

describe("Login Lambda", () => {
  beforeEach(() => {
    vi.resetAllMocks();
  });

  it("handles a registration request with email, password and sports", async () => {
    dynamoDbSend.mockReturnValue({ Item: [] });
    const testEvent: APIGatewayProxyEvent = getTestEvent({
      email: "my@test.com",
      password: "asdf",
      sports: ["Vollyball", "Basketball"],
    });

    const result = await handler(testEvent);

    expect(identityProviderSend).toHaveBeenCalled();

    expect(dynamoDbSend).toHaveBeenCalled();

    expect(result).toMatchSnapshot();
  });

  it("handles a registration request with email, password and sports as base64", async () => {
    dynamoDbSend.mockReturnValue({ Item: [] });
    const testEvent: APIGatewayProxyEvent = getTestEvent(
      {
        email: "my@test.com",
        password: "asdf",
        sports: ["Vollyball", "Basketball"],
      },
      true,
    );

    const result = await handler(testEvent);

    expect(identityProviderSend).toHaveBeenCalled();

    expect(dynamoDbSend).toHaveBeenCalled();

    expect(result).toMatchSnapshot();
  });

  it.each`
    email            | password     | sports           | expected
    ${undefined}     | ${"qwert1T"} | ${["Vollyball"]} | ${400}
    ${"my@test.com"} | ${undefined} | ${["Vollyball"]} | ${400}
    ${""}            | ${"qwert1T"} | ${["Vollyball"]} | ${400}
    ${"my@test.com"} | ${""}        | ${["Vollyball"]} | ${400}
    ${""}            | ${""}        | ${["Vollyball"]} | ${400}
    ${undefined}     | ${undefined} | ${["Vollyball"]} | ${400}
    ${"my@test.com"} | ${"qwert1T"} | ${[]}            | ${400}
    ${"my@test.com"} | ${"qwert1T"} | ${undefined}     | ${400}
    ${"my@test.com"} | ${"qwert1T"} | ${"Vollyball"}   | ${400}
  `(
    "handles a login request with incomplete payload",
    async ({ email, password, expected }) => {
      dynamoDbSend.mockReturnValue({ Item: [] });
      const testEvent: APIGatewayProxyEvent = getTestEvent({
        email,
        password,
      });

      const result = await handler(testEvent);

      expect(identityProviderSend).not.toHaveBeenCalled();

      expect(dynamoDbSend).not.toHaveBeenCalled();

      expect(result).toMatchSnapshot();
    },
  );

  it("rejects an unknown user", async () => {
    const testEvent: APIGatewayProxyEvent = getTestEvent({
      email: "unknown@unknown.de",
      password: "unkndown",
      sports: ["Vollyball"],
    });

    const error = {
      name: "UserNotFoundException",
      $fault: "client",
      $metadata: {
        httpStatusCode: 400,
        requestId: "3fccc512-2371-4782-95e5-94b01d2b29ae",
        attempts: 1,
        totalRetryDelay: 0,
      },
      __type: "UserNotFoundException",
    };

    identityProviderSend.mockImplementationOnce(() => {
      throw error;
    });

    const result = await handler(testEvent);

    expect(identityProviderSend).toHaveBeenCalled();

    expect(dynamoDbSend).not.toHaveBeenCalled();

    expect(result).toMatchSnapshot();
  });
});
