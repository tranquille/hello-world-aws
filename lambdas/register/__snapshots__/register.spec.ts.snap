// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login Lambda > handles a login request with incomplete payload 1`] = `
{
  "body": "\\"provided payload is not valid\\"",
  "headers": {
    "Content-Type": "application/json",
  },
  "isBase64Encoded": false,
  "statusCode": 400,
}
`;

exports[`Login Lambda > handles a login request with incomplete payload 2`] = `
{
  "body": "\\"provided payload is not valid\\"",
  "headers": {
    "Content-Type": "application/json",
  },
  "isBase64Encoded": false,
  "statusCode": 400,
}
`;

exports[`Login Lambda > handles a login request with incomplete payload 3`] = `
{
  "body": "\\"provided payload is not valid\\"",
  "headers": {
    "Content-Type": "application/json",
  },
  "isBase64Encoded": false,
  "statusCode": 400,
}
`;

exports[`Login Lambda > handles a login request with incomplete payload 4`] = `
{
  "body": "\\"provided payload is not valid\\"",
  "headers": {
    "Content-Type": "application/json",
  },
  "isBase64Encoded": false,
  "statusCode": 400,
}
`;

exports[`Login Lambda > handles a login request with incomplete payload 5`] = `
{
  "body": "\\"provided payload is not valid\\"",
  "headers": {
    "Content-Type": "application/json",
  },
  "isBase64Encoded": false,
  "statusCode": 400,
}
`;

exports[`Login Lambda > handles a login request with incomplete payload 6`] = `
{
  "body": "\\"provided payload is not valid\\"",
  "headers": {
    "Content-Type": "application/json",
  },
  "isBase64Encoded": false,
  "statusCode": 400,
}
`;

exports[`Login Lambda > handles a login request with incomplete payload 7`] = `
{
  "body": "\\"provided payload is not valid\\"",
  "headers": {
    "Content-Type": "application/json",
  },
  "isBase64Encoded": false,
  "statusCode": 400,
}
`;

exports[`Login Lambda > handles a login request with incomplete payload 8`] = `
{
  "body": "\\"provided payload is not valid\\"",
  "headers": {
    "Content-Type": "application/json",
  },
  "isBase64Encoded": false,
  "statusCode": 400,
}
`;

exports[`Login Lambda > handles a login request with incomplete payload 9`] = `
{
  "body": "\\"provided payload is not valid\\"",
  "headers": {
    "Content-Type": "application/json",
  },
  "isBase64Encoded": false,
  "statusCode": 400,
}
`;

exports[`Login Lambda > handles a registration request with email, password and sports 1`] = `
{
  "body": undefined,
  "headers": {
    "Content-Type": "application/json",
  },
  "isBase64Encoded": false,
  "statusCode": 201,
}
`;

exports[`Login Lambda > handles a registration request with email, password and sports as base64 1`] = `
{
  "body": undefined,
  "headers": {
    "Content-Type": "application/json",
  },
  "isBase64Encoded": false,
  "statusCode": 201,
}
`;

exports[`Login Lambda > rejects an unknown user 1`] = `
{
  "body": "{\\"name\\":\\"UserNotFoundException\\",\\"$fault\\":\\"client\\",\\"$metadata\\":{\\"httpStatusCode\\":400,\\"requestId\\":\\"3fccc512-2371-4782-95e5-94b01d2b29ae\\",\\"attempts\\":1,\\"totalRetryDelay\\":0},\\"__type\\":\\"UserNotFoundException\\"}",
  "headers": {
    "Content-Type": "application/json",
  },
  "isBase64Encoded": false,
  "statusCode": 400,
}
`;
